pkg/controllers/pod/podController.go

```diff
@@ -0,0 +1,314 @@
+func (c *PodController) Run(threadiness int, reconcilerPeriod string, stopCh chan struct{}) {
+	defer uruntime.HandleCrash()
+
+	// Let the workers stop when we are done
+	workqueue := c.labelCache.GetQueue()
+	defer workqueue.ShutDown()
+
+	log.Info("Starting pod controller")
+
+	// Start Calico cache. Cache gets loaded with objects
+	// from ETCD datastore.
+	c.labelCache.Run(reconcilerPeriod)
+
+	// Load endpoint cache. Retry when failed.
+	for err := c.populateEndpointCache(); err != nil; {
+		log.WithError(err).Errorf("Failed to load workload endpoint cache, retrying")
+	}
+
+	go c.informer.Run(stopCh)
+
+	// Wait till k8s cache is synced
+	for !c.informer.HasSynced() {
+	}
+
+	// Start a number of worker threads to read from the queue.
+	for i := 0; i < threadiness; i++ {
+		go c.runWorker()
+	}
+
+	<-stopCh
+	log.Info("Stopping Pod controller")
+}
+
+func (c *PodController) runWorker() {
+	for c.processNextItem() {
+	}
+}
+
+func (c *PodController) processNextItem() bool {
+	// Wait until there is a new item in the working queue
+	workqueue := c.labelCache.GetQueue()
+	key, quit := workqueue.Get()
+	if quit {
+		return false
+	}
+	// Tell the queue that we are done with processing this key. This unblocks the key for other workers
+	// This allows safe parallel processing because two nodes with the same key are never processed in
+	// parallel.
+	defer workqueue.Done(key)
+
+	// Invoke the method containing the business logic
+	err := c.syncToCalico(key.(string))
+
+	// Handle the error if something went wrong during the execution of the business logic
+	c.handleErr(err, key.(string))
+	return true
+}
+
+// syncToCalico syncs the given update to Calico's etcd
+func (c *PodController) syncToCalico(key string) error {
+
+	// Check if it exists in our cache.
+	if labels, exists := c.labelCache.Get(key); exists {
+
+		// Get workloadEndpoint from cache
+		endpoint, exists := c.endpointCache[key]
+
+		if !exists {
+			log.Errorf("No endpoint for %s in cache, loading", key)
+
+			// Load endpoint cache. Retry when failed.
+			for err := c.populateEndpointCache(); err != nil; {
+				log.WithError(err).Errorf("Failed to load workload endpoint cache, retrying.")
+			}
+
+			endpoint, exists = c.endpointCache[key]
+			if !exists {
+				// No endpoint in etcd - this means the pod hasn't been
+				// created by the CNI plugin yet. Just wait until it has been.
+				// This can only be hit when labels for a pod change before
+				//  the pod has been deployed, so should be pretty uncommon.
+				log.Errorf("Pod %s hasn't been created by the CNI plugin yet.", key)
+				return nil
+			}
+		}
+
+		oldLabels := endpoint.Metadata.Labels
+		newLabels := labels.(map[string]string)
+
+		if !reflect.DeepEqual(oldLabels, newLabels) {
+
+			// Update the labels on endpoint
+			endpoint.Metadata.Labels = newLabels
+
+			log.Infof("Applying endpoint %s on ETCD \n", key)
+			_, applyErr := c.calicoClient.WorkloadEndpoints().Apply(&endpoint)
```

Comment:
I think we should never be doing an `Apply` on an endpoint - we should only ever do an `Update`. 

That way we'll never accidentally create a WorkloadEndpoint that was deleted by the CNI plugin already.  We'll need to handle the "does not exist" error.